SPARQL Repository: Not exhausting resultset may cause problems on closing resultset with try-with-resource

Test environment was a local RDF4J server. Version 2.2.1 The server containend ~ 100000 triples

The following code fails after the second or third run with a socket closed exception:

```
try(RepositoryConnection repCon = RDF4JSparqlRepo().getConnection() ) {
            for(int i = 0; i < 10; i++) {
                TupleQuery entityQuery = repCon.prepareTupleQuery(QueryLanguage.SPARQL, "select * where {?s ?p ?o} limit 2000");

                try (TupleQueryResult result = entityQuery.evaluate()) {
                    result.next();
                } //fails here on close
            }
        }
```

The code works if I exhaust the resultset before:

```
try(RepositoryConnection repCon = RDF4JSparqlRepo().getConnection()) {
            for(int i = 0; i < 10; i++) {
                TupleQuery entityQuery = repCon.prepareTupleQuery(QueryLanguage.SPARQL, "select * where {?s ?p ?o} limit 2000");

                try (TupleQueryResult result = entityQuery.evaluate()) {
                    while(result.hasNext()) {
                        result.next();
                    }
                }
            }
        }

```

Errors are various e.g.

Caused by: org.eclipse.rdf4j.query.QueryEvaluationException: java.net.SocketException: Socket closed
	at org.eclipse.rdf4j.query.resultio.helpers.BackgroundTupleResult.handleClose(BackgroundTupleResult.java:82)
	at org.eclipse.rdf4j.common.iteration.AbstractCloseableIteration.close(AbstractCloseableIteration.java:51)
	at Main.main(Main.java:39)
	
Caused by: java.lang.ArrayIndexOutOfBoundsException
    at java.net.SocketInputStream.read(SocketInputStream.java:162)
    at java.net.SocketInputStream.read(SocketInputStream.java:141)
    at org.apache.http.impl.io.SessionInputBufferImpl.streamRead(SessionInputBufferImpl.java:139)
	
	
Caused by: org.apache.http.MalformedChunkCodingException: Unexpected content at the end of chunk
	at org.apache.http.impl.io.ChunkedInputStream.getChunkSize(ChunkedInputStream.java:259)
	at org.apache.http.impl.io.ChunkedInputStream.nextChunk(ChunkedInputStream.java:227)
	at org.apache.http.impl.io.ChunkedInputStream.read(ChunkedInputStream.java:186)
	at org.apache.http.impl.io.ChunkedInputStream.read(ChunkedInputStream.java:215)

This may be the same issue that was reported here https://github.com/eclipse/rdf4j/issues/853

--- a/core/query/src/main/java/org/eclipse/rdf4j/query/impl/BackgroundGraphResult.java
+++ b/core/query/src/main/java/org/eclipse/rdf4j/query/impl/BackgroundGraphResult.java
@@ -9,11 +9,9 @@ package org.eclipse.rdf4j.query.impl;
 
 import java.io.InputStream;
 import java.io.InputStreamReader;
-import java.lang.reflect.UndeclaredThrowableException;
 import java.nio.charset.Charset;
 import java.util.Collections;
 import java.util.Map;
-import java.util.NoSuchElementException;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.CountDownLatch;
 
@@ -45,6 +43,8 @@ public class BackgroundGraphResult extends IterationWrapper<Statement, QueryEval
 
 	private final CountDownLatch namespacesReady = new CountDownLatch(1);
 
+	private final CountDownLatch finishedParsing = new CountDownLatch(1);
+
 	private final Map<String, String> namespaces = new ConcurrentHashMap<String, String>();
 
 	private final QueueCursor<Statement> queue;
@@ -65,67 +65,6 @@ public class BackgroundGraphResult extends IterationWrapper<Statement, QueryEval
 	}
 
 	@Override
-	public boolean hasNext()
-		throws QueryEvaluationException
-	{
-		if (isClosed()) {
-			return false;
-		}
-		if (Thread.currentThread().isInterrupted()) {
-			close();
-			return false;
-		}
-
-		boolean result = queue.hasNext();
-		if (!result) {
-			close();
-		}
-		return result;
-	}
-
-	@Override
-	public Statement next()
-		throws QueryEvaluationException
-	{
-		if (isClosed()) {
-			throw new NoSuchElementException("The iteration has been closed.");
-		}
-		if (Thread.currentThread().isInterrupted()) {
-			close();
-			throw new NoSuchElementException("The iteration has been closed.");
-		}
-
-		try {
-			return queue.next();
-		}
-		catch (NoSuchElementException e) {
-			close();
-			throw e;
-		}
-	}
-
-	@Override
-	public void remove()
-		throws QueryEvaluationException
-	{
-		if (isClosed()) {
-			throw new IllegalStateException("The iteration has been closed.");
-		}
-		if (Thread.currentThread().isInterrupted()) {
-			close();
-			throw new IllegalStateException("The iteration has been closed.");
-		}
-
-		try {
-			queue.remove();
-		}
-		catch (IllegalStateException e) {
-			close();
-			throw e;
-		}
-	}
-
-	@Override
 	protected void handleClose()
 		throws QueryEvaluationException
 	{
@@ -135,30 +74,38 @@ public class BackgroundGraphResult extends IterationWrapper<Statement, QueryEval
 		finally {
 			queue.done();
 		}
+		try {
+			finishedParsing.await();
+		}
+		catch (InterruptedException e) {
+			Thread.currentThread().interrupt();
+		} finally {
+			queue.checkException();
+		}
 	}
 
 	@Override
 	public void run() {
 		try {
-			parser.setRDFHandler(this);
-			if (charset == null) {
-				parser.parse(in, baseURI);
+			try {
+				parser.setRDFHandler(this);
+				if (charset == null) {
+					parser.parse(in, baseURI);
+				}
+				else {
+					parser.parse(new InputStreamReader(in, charset), baseURI);
+				}
+			} finally {
+				in.close();
 			}
-			else {
-				parser.parse(new InputStreamReader(in, charset), baseURI);
-			}
-		}
-		catch (RDFHandlerException e) {
-			// parsing was cancelled or interrupted
-			close();
 		}
 		catch (Exception e) {
 			queue.toss(e);
-			close();
 		}
 		finally {
 			queue.done();
 			namespacesReady.countDown();
+			finishedParsing.countDown();
 		}
 	}
 
@@ -178,8 +125,9 @@ public class BackgroundGraphResult extends IterationWrapper<Statement, QueryEval
 		}
 		catch (InterruptedException e) {
 			Thread.currentThread().interrupt();
-			close();
-			throw new UndeclaredThrowableException(e);
+			return Collections.emptyMap();
+		} finally {
+			queue.checkException();
 		}
 	}
 
@@ -207,11 +155,8 @@ public class BackgroundGraphResult extends IterationWrapper<Statement, QueryEval
 		}
 		catch (InterruptedException e) {
 			Thread.currentThread().interrupt();
-			close();
-			throw new RDFHandlerException(e);
-		}
-		if (isClosed()) {
-			throw new RDFHandlerException("Result closed");
+			queue.toss(e);
+			queue.done();
 		}
 	}
 

--- a/core/query/src/main/java/org/eclipse/rdf4j/query/impl/BackgroundGraphResult.java
+++ b/core/query/src/main/java/org/eclipse/rdf4j/query/impl/BackgroundGraphResult.java
@@ -7,7 +7,6 @@
  *******************************************************************************/
 package org.eclipse.rdf4j.query.impl;
 
-import java.io.IOException;
 import java.io.InputStream;
 import java.io.InputStreamReader;
 import java.lang.reflect.UndeclaredThrowableException;
@@ -131,28 +130,10 @@ public class BackgroundGraphResult extends IterationWrapper<Statement, QueryEval
 		throws QueryEvaluationException
 	{
 		try {
-			try {
-				super.handleClose();
-			}
-			finally {
-				try {
-					// After checking that we ourselves cannot possibly be generating an NPE ourselves, 
-					// attempt to close the input stream we were given
-					InputStream toClose = in;
-					if (toClose != null) {
-						toClose.close();
-					}
-				}
-				catch (NullPointerException e) {
-					// Swallow NullPointerException that Apache HTTPClient is hiding behind a NotThreadSafe annotation
-				}
-			}
-		}
-		catch (IOException e) {
-			throw new QueryEvaluationException(e);
+			super.handleClose();
 		}
 		finally {
-			queue.close();
+			queue.done();
 		}
 	}
 

--- a/core/queryresultio/api/src/main/java/org/eclipse/rdf4j/query/resultio/helpers/BackgroundTupleResult.java
+++ b/core/queryresultio/api/src/main/java/org/eclipse/rdf4j/query/resultio/helpers/BackgroundTupleResult.java
@@ -7,7 +7,6 @@
  *******************************************************************************/
 package org.eclipse.rdf4j.query.resultio.helpers;
 
-import java.io.IOException;
 import java.io.InputStream;
 import java.lang.reflect.UndeclaredThrowableException;
 import java.util.ArrayList;
@@ -61,28 +60,10 @@ public class BackgroundTupleResult extends IteratingTupleQueryResult
 		throws QueryEvaluationException
 	{
 		try {
-			try {
-				super.handleClose();
-			}
-			finally {
-				try {
-					// After checking that we ourselves cannot possibly be generating an NPE ourselves, 
-					// attempt to close the input stream we were given
-					InputStream toClose = in;
-					if (toClose != null) {
-						toClose.close();
-					}
-				}
-				catch (NullPointerException e) {
-					// Swallow NullPointerException that Apache HTTPClient is hiding behind a NotThreadSafe annotation
-				}
-			}
-		}
-		catch (IOException e) {
-			throw new QueryEvaluationException(e);
+			super.handleClose();
 		}
 		finally {
-			queue.close();
+			queue.done();
 		}
 	}
 

--- a/core/queryresultio/api/src/main/java/org/eclipse/rdf4j/query/resultio/helpers/BackgroundTupleResult.java
+++ b/core/queryresultio/api/src/main/java/org/eclipse/rdf4j/query/resultio/helpers/BackgroundTupleResult.java
@@ -8,7 +8,6 @@
 package org.eclipse.rdf4j.query.resultio.helpers;
 
 import java.io.InputStream;
-import java.lang.reflect.UndeclaredThrowableException;
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.List;
@@ -42,6 +41,8 @@ public class BackgroundTupleResult extends IteratingTupleQueryResult
 
 	private final CountDownLatch bindingNamesReady = new CountDownLatch(1);
 
+	private final CountDownLatch finishedParsing = new CountDownLatch(1);
+
 	public BackgroundTupleResult(TupleQueryResultParser parser, InputStream in) {
 		this(new QueueCursor<BindingSet>(10), parser, in);
 	}
@@ -65,43 +66,47 @@ public class BackgroundTupleResult extends IteratingTupleQueryResult
 		finally {
 			queue.done();
 		}
+		try {
+			finishedParsing.await();
+		}
+		catch (InterruptedException e) {
+			Thread.currentThread().interrupt();
+		} finally {
+			queue.checkException();
+		}
 	}
 
 	@Override
 	public List<String> getBindingNames() {
 		try {
 			bindingNamesReady.await();
-			queue.checkException();
 			return bindingNames;
 		}
 		catch (InterruptedException e) {
 			Thread.currentThread().interrupt();
-			close();
-			throw new UndeclaredThrowableException(e);
-		}
-		catch (QueryEvaluationException e) {
-			close();
-			throw new UndeclaredThrowableException(e);
+			return Collections.emptyList();
+		} finally {
+			queue.checkException();
 		}
 	}
 
 	@Override
 	public void run() {
 		try {
-			parser.setQueryResultHandler(this);
-			parser.parseQueryResult(in);
-		}
-		catch (QueryResultHandlerException e) {
-			// parsing cancelled or interrupted
-			close();
+			try {
+				parser.setQueryResultHandler(this);
+				parser.parseQueryResult(in);
+			} finally {
+				in.close();
+			}
 		}
 		catch (Exception e) {
 			queue.toss(e);
-			close();
 		}
 		finally {
 			queue.done();
 			bindingNamesReady.countDown();
+			finishedParsing.countDown();
 		}
 	}
 
@@ -122,11 +127,8 @@ public class BackgroundTupleResult extends IteratingTupleQueryResult
 		}
 		catch (InterruptedException e) {
 			Thread.currentThread().interrupt();
-			close();
-			throw new TupleQueryResultHandlerException(e);
-		}
-		if (isClosed()) {
-			throw new TupleQueryResultHandlerException("Result closed");
+			queue.toss(e);
+			queue.done();
 		}
 	}
 

--- a/core/util/src/main/java/org/eclipse/rdf4j/common/iteration/IterationWrapper.java
+++ b/core/util/src/main/java/org/eclipse/rdf4j/common/iteration/IterationWrapper.java
@@ -60,6 +60,10 @@ public class IterationWrapper<E, X extends Exception> extends AbstractCloseableI
 		if (isClosed()) {
 			return false;
 		}
+		else if (Thread.currentThread().isInterrupted()) {
+			close();
+			return false;
+		}
 		boolean result = wrappedIter.hasNext();
 		if (!result) {
 			close();
@@ -78,6 +82,10 @@ public class IterationWrapper<E, X extends Exception> extends AbstractCloseableI
 	{
 		if (isClosed()) {
 			throw new NoSuchElementException("The iteration has been closed.");
+		}
+		else if (Thread.currentThread().isInterrupted()) {
+			close();
+			throw new NoSuchElementException("The iteration has been interrupted.");
 		}
 		try {
 			return wrappedIter.next();
@@ -102,6 +110,10 @@ public class IterationWrapper<E, X extends Exception> extends AbstractCloseableI
 	{
 		if (isClosed()) {
 			throw new IllegalStateException("The iteration has been closed.");
+		}
+		else if (Thread.currentThread().isInterrupted()) {
+			close();
+			throw new IllegalStateException("The iteration has been interrupted.");
 		}
 		try {
 			wrappedIter.remove();

--- a/core/util/src/main/java/org/eclipse/rdf4j/common/iteration/QueueIteration.java
+++ b/core/util/src/main/java/org/eclipse/rdf4j/common/iteration/QueueIteration.java
@@ -142,8 +142,8 @@ public abstract class QueueIteration<E, T extends Exception> extends LookAheadIt
 				}
 			}
 			if (isAfterLast(take)) {
-				checkException();
 				done(); // put afterLast back for others
+				checkException();
 				return null;
 			}
 			checkException();

